Upgrade file picker app buttons

LMS007 opened the pull request:
~depends on https://github.com/hypothesis/frontend-shared/pull/48~
~depends on https://github.com/hypothesis/frontend-shared/pull/67~

### Commits
-------

Remove Button.js and corresponding scss

-------

Update buttons in FilePickerApp 

- Switch Button over to LabeledButton in FilePickerApp
- Rename several classes in FilePickerApp
- Adjust css and html of button wrapper to work as a column lay out rather than use inline-flex
- TODO: Replace FilePickerApp__actions* with exported classes from frontend-shared when ready.

-------

Replace data-test with data-testid 

This matches our convention for the testing id attribute

--------

### Discussion 

**Before**
![May-05-2021 17-07-13](https://user-images.githubusercontent.com/3939074/117224301-713dc180-adc4-11eb-92bc-cde2affefd37.gif)

**After**
![Screen Shot 2021-05-18 at 9 18 27 AM](https://user-images.githubusercontent.com/3939074/118687575-0922ae80-b7ba-11eb-8ec8-1190e9cfbc16.png)


Biggest visual difference is the button is not as dark (which matches are primary color) and there is a transition effect on hover. The padding is also a bit different and there is no more red hover color--this matches other LabeledButtons.

lyzadanger reviewed (changes requested):
Hiya! I think this can be accomplished with less customization—after a bit of experimentation, I think the left-alignment of the button text may be the only thing we need to override on the button itself. See what you think.

* I don't think the hover should be red (unless there is a compelling reason other than "it was red before") as this doesn't match our established button patterns.
* Several of the rules currently applied to the customized button class can be tackled in the parent element (more on this shortly).
* As the `layout-space` variables and handling of same are currently being re-established and re-considered in `frontend-shared`, let's avoid copying them around at this time.

Regarding some of the layout styling we're trying to achieve with these buttons:

* You can use a `flex-column` on a containing element to lay out the buttons in the columnar presentation we're aiming for here. The `.u-layout-column` utility style (from the `client` and just about to land in `frontend-shared`) has rules that do what you'd need:

```css
.button-parent-container-whatever {
    display: flex;
    flex-direction: column;
    justify-content: flex-start;
    align-items: stretch;
}
```

The width of the flex column container will establish the width of the buttons (they will all stretch to fill).
* Establish the vertical spacing between the buttons on the flex container element (not on the buttons themselves. Our general convention is that elements/components should not assert their own margins—that's the job of the parent container. See the client's `vertical-rhythm` mixin/utility style for implementation. 
* I can see you're overriding padding but I'm unsure why. You can use the `variant="large"` prop value to get a little more padding by default, which will match the established patterns. 

This leaves the alignment of text in the buttons, which may be a legitimate need for customization (`justify-content: flex-start;`).

lyzadanger commented:
Also: it looks like the cited dependency is merged. Can you update the PR description if so?

robertknight commented:
FYI a new release of the @hypothesis/frontend-shared package has been published and `LabeledButton` now sets `type="button"` by default. An issue has also been filed related to supporting [control over label alignment](https://github.com/hypothesis/frontend-shared/issues/63).

LMS007 commented on lms/static/styles/components/_FilePickerApp.scss:
```diff
@@ -32,8 +29,9 @@
 }
 
 .FilePickerApp__document-source-buttons {
```
I could have also added the vertical rhythm mixin here, but since it was unchanged from the util class I opted to include the actual exported class in the component directly. "u-vertical-rhythm"

lyzadanger replied:
Danger zone! Recall from this PR yesterday: https://github.com/hypothesis/frontend-shared/pull/66 that all utility classes are about to get a prefix. Also, there is documentation here: https://github.com/hypothesis/frontend-shared/tree/main/styles about what the current package SASS API is, and (in the FAQ) section, specifically advises against using any utility classes directly. In short, this class won't exist after the next package release.

lyzadanger replied:
See comment further down: this will break on the next `frontend-shared` package release.

LMS007 commented on lms/static/styles/components/_FilePickerApp.scss:
```diff
@@ -32,8 +29,9 @@
 
 .FilePickerApp__document-source-buttons {
-  align-items: stretch;
+  @include layout.column($align: 'flex-start');
```
This endined being one way to keep the buttons from going full width. I had trouble if the display was set to "flex", I do wonder if there is an even simpler or better approach?

LMS007 commented on lms/static/styles/components/_FilePickerApp.scss:
```diff
@@ -32,8 +29,9 @@
+  @include layout.column($align: 'flex-start');
+  // Override layout.column's display from "flex" to "inline-flex" so the buttons don't
+  // take the full width
   display: inline-flex;
```
This ended up being one way to keep the buttons from going full width. I had trouble if the display was set to "flex", I do wonder if there is an even simpler or better approach?

lyzadanger replied:
My gut hunch is that there probably is a more straightforward way, but I am not certain: I'd have to get in and poke around.

One suggestion: perhaps don't try to use any mixins or utilities and just see what CSS is needed to get the job done. A columnar flex layout does seem plausible...But the `inline-flex` display gives me pause.

LMS007 replied:
After some careful consideration, I think what is here currently seems reasonable. We have a flex container set to vertical, but we don't want to take the full width, so we use inline-flex. Button items then flow and fill up the space that the longest one takes.

LMS007 commented on lms/static/styles/frontend_apps.scss:
```diff
@@ -30,6 +30,7 @@
 
 // Shared styles from frontend-shared package
 @use '@hypothesis/frontend-shared/styles';
+@use '@hypothesis/frontend-shared/styles/util';
```
Part of my thinks we should get all the exported classes in one go, this feels a bit confusing, but I realize it's still a new package.

lyzadanger replied:
Danger zone! Recall from this PR yesterday: https://github.com/hypothesis/frontend-shared/pull/66 that all utility classes are about to get a prefix. Also, there is documentation here: https://github.com/hypothesis/frontend-shared/tree/main/styles about what the current package SASS API is, and (in the FAQ) section, specifically advises against using any utility classes directly. In short, this class won't exist after the next package release.

lyzadanger commented on lms/static/styles/components/_FilePickerApp.scss:
```diff
@@ -1,3 +1,6 @@
+@use '@hypothesis/frontend-shared/styles/mixins/layout';
+@use '@hypothesis/frontend-shared/styles/components/buttons';
```
Is this used?

LMS007 replied:
left over vestige from previous effort, good catch.

lyzadanger reviewed (changes requested):
Hiya! Alas, using any of the shared-package utility CSS classes is not currently feasible (see my inline comments). Feel free to ping me for a voice-chat or Slack if you want to run other approaches by me as you move forward.

robertknight commented:
As part of https://github.com/hypothesis/lms/pull/2687 I'm planning to extract the part of `FilePickerApp` that deals with selecting content into a separate component. That will end up obsoleting the changes here. The issue in https://github.com/hypothesis/frontend-shared/issues/63 remains valid, so I would suggest resolving that first.

LMS007 commented:
Updated PR top depend on button align fix in FES. I also adjusted the vertical spacing to match the soon-to-land column rule where it uses 1/2 REM rather than 1 REM. 

I was not able to replicate `hyp-actions--column` 
e.g.

```
@include layout.column;    
@include layout.vertical-rhythm($-space * 0.5);
```

here because that allows the container to expand the full width. In this case it's the form with is 600px.

LMS007 commented:
Converting to draft until hypothesis/frontend-shared#67 lands

lyzadanger commented:
> I also adjusted the vertical spacing to match the soon-to-land column rule where it uses 1/2 REM rather than 1 REM.

Please don't! It's dangerous to start suddenly using `rem` units in one place without looking at the bigger picture.

lyzadanger commented on lms/static/styles/components/_FilePickerApp.scss:
```diff
@@ -1,18 +1,16 @@
+@use '@hypothesis/frontend-shared/styles/mixins/layout';
```
I think this needs to be removed.

robertknight commented:
I have [opened a PR](https://github.com/hypothesis/lms/pull/2707) that extracts the content selection UI/logic out of `FilePickerApp` into a separate component as part of https://github.com/hypothesis/lms/issues/2524.

That PR conflicts with this one, so if this lands first we'll need to copy across the changes into that PR.

lyzadanger commented:
I have pushed a fixup commit: https://github.com/hypothesis/lms/pull/2646/commits/58eb5f0b12312801b902c9098b6779954adeecec that addresses HTML (containers) and styling for the LMS File Picker action buttons.

As of this commit, the file-picker buttons look like this:

<img width="566" alt="Screen Shot 2021-05-19 at 12 30 15 PM" src="https://user-images.githubusercontent.com/439947/118851667-039f9400-b8a0-11eb-9b00-836024b39990.png">

The left alignment will come when the relevant change is released in `frontend-shared`.

I have not resolved the (pre-existing) conflict between this branch and `master`—please let me know if I'm needed for that task.

I'm going to un-assign myself as reviewer for this PR.

robertknight commented on lms/static/styles/components/_FilePickerApp.scss:
```diff
@@ -31,9 +27,26 @@
+.FilePickerApp__actions {
+  // A flex container (row) with two children: a column of buttons and a
+  // "spacer" to take up the rest of the row space not needed by the buttons.
+  // The spacer prevents the button container from stretching full width.
```
I think "stretching full width" is meant to say "stretching full height"?

LMS007 replied:
The spacer is for width

```
Container text which is longer than the longest button
[button 1            ] [spacer.............................................
[b u t t o n 2       ]  
[b  u  t  t  o  n  3 ]                                                      ]
```

robertknight replied:
Oh, right. I misread the first line that said "flex container (row)".

robertknight commented on lms/static/styles/components/_FilePickerApp.scss:
```diff
@@ -1,18 +1,14 @@
 @use "../variables" as var;
 
+$-space: 1rem; // TODO: replace this with a real variable
```
This variable is unused.

LMS007 replied:
aw.. missed that!

robertknight commented on lms/static/styles/components/_FilePickerApp.scss:
```diff
@@ -31,9 +27,26 @@
+    // Display buttons in a column with a vertical margin between each button
+    display: flex;
+    flex-direction: column;
+    & > * + * {
```
A possible way to achieve the same result without `!important`:

```scss
& > :not(:first-child) {
  margin-top: 0.5em;
}
```

lyzadanger replied:
This is a very interesting comment, thank you. A simple way to bump up specificity. Might I suggest:

* We leave this as-is for the moment, as the `!important` is caged in pretty well and is probably _safe_, and
* I have an issue where I'm tracking some work related to the existing layout utility classes and mixins: https://github.com/hypothesis/frontend-shared/issues/74 . This selector/rule is based on the `vertical-rhythm` mixin/pattern. I could evaluate at that time using this higher-specificity strategy for those mixins and utility classes.

In summary, I expect these new rules (here, in this LMS component's SASS) to have a short lifespan, to be (aspirationally) replaced by shared patterns soon.

LMS007 replied:
Because this is going to be replaced shortly, I'm in favor of just leaving this. We may want to include Rob's suggestion in the util class in the shared  lib.

robertknight commented on lms/static/styles/components/_FilePickerApp.scss:
```diff
@@ -29,9 +27,24 @@ $-space: 1rem; // TODO: replace this with a real variable
+    }
+  }
+
+  &-spacer {
```
Since this utility isn't doing anything besides `u-stretch` using the utility class would save a little code here. If a bunch of other rules were added to the selector then I think that would better justify a new class.

lyzadanger replied:
What I was trying to get at in my comment is that I acknowledge that the `.u-stretch` class would have accomplished the same thing, but felt that ergonomically for the time being it was easier to understand if all of the rules were in one place in this particular case. It's easier to look at as a whole and squint and think: "Is this a pattern that should/can be extracted?" I see this styling as temporary, in that I hope we'll have a reusable columnar-actions pattern out of `frontend-shared` in the near- to medium-term. (Such a pattern already exists in `frontend-shared`, but it's still in flux as to how we'd provide this as part of the public SASS/CSS API).

Anyway, if you feel strongly, we can use `.u-stretch`. I'm not that passionate!

robertknight commented on lms/static/scripts/frontend_apps/components/test/FilePickerApp-test.js:
```diff
@@ -100,35 +100,35 @@ describe('FilePickerApp', () => {
     fakeConfig.filePicker.canvas.enabled = true;
     const wrapper = renderFilePicker();
-    assert.isTrue(wrapper.exists('Button[label="Select PDF from Canvas"]'));
+    assert.isTrue(wrapper.exists('LabeledButton[data-testid="pdf-button"]'));
```
👍 - Changing label matching to `data-testid` seems good here. We'll need to do this more widely when we add support for other languages which might happen later this year.

robertknight reviewed (commented):
Visually this looks good. The change to matching based on `data-testid`s rather than labels in the tests is I think a good change. I had some minor comments on the code:

- I found one unused SASS variable
- I had a couple of suggestions regarding styling (the `margin-top` override and use of `u-stretch`)

lyzadanger commented:
@LMS007 @robertknight 

I've responded to a couple of @robertknight 's comments pertaining to SASS rules and strategy. I'm not going to make any code changes at this time and will leave it to @LMS007 to remove the unused SASS variable and any other changes needed to land this!

robertknight commented:
> Lyza fixed her issues herself

That was poorly worded. What I meant is that Lyza addressed the issues she raised in an earlier review by making changes directly to the code.
